pkg/kube/controllers/boshdeployment/reconciler.go

```diff
@@ -83,50 +128,345 @@ func (r *ReconcileBOSHDeployment) Reconcile(request reconcile.Request) (reconcil
+func (r *ReconcileBOSHDeployment) generateVariables(ctx context.Context, instance *bdc.BOSHDeployment, manifest *bdm.Manifest, kubeConfig *bdm.KubeConfig) error {
+	r.log.Debug("Creating variables extendedSecrets")
+	var err error
+	for _, variable := range kubeConfig.Variables {
+		// Set BOSHDeployment instance as the owner and controller
+		if err := r.setReference(instance, &variable, r.scheme); err != nil {
+			r.recorder.Event(instance, corev1.EventTypeWarning, "NewExtendedStatefulSetForDeployment Error", err.Error())
+			return errors.Wrap(err, "couldn't set reference for an ExtendedStatefulSet for a BOSH Deployment")
+		}
+
+		foundSecret := &esv1.ExtendedSecret{}
+		err = r.client.Get(ctx, types.NamespacedName{Name: variable.GetName(), Namespace: variable.GetNamespace()}, foundSecret)
+		if apierrors.IsNotFound(err) {
+			err = r.client.Create(ctx, &variable)
+			if err != nil {
+				return errors.Wrapf(err, "could not create ExtendedSecret %s", variable.GetName())
+			}
+		} else {
+			foundSecret.Spec = variable.Spec
+			err = r.client.Update(ctx, foundSecret)
+			if err != nil {
+				return errors.Wrapf(err, "could not update ExtendedSecret %s", variable.GetName())
+			}
+		}
+	}
+
+	instance.Status.State = variableGeneratedState
+
+	return nil
+}
+
+// createVariableInterpolationExJob create temp manifest and variable interpolation exJob
+func (r *ReconcileBOSHDeployment) createVariableInterpolationExJob(ctx context.Context, instance *bdc.BOSHDeployment, manifest *bdm.Manifest, variables []esv1.ExtendedSecret, jobLabels map[string]string, secretLabels map[string]string) error {
+	// Create temp manifest as variable interpolation job input
+	tempManifestBytes, err := yaml.Marshal(manifest)
+	if err != nil {
+		return errors.Wrap(err, "could not marshal temp manifest")
+	}
+
+	tempManifestSecret := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      manifest.GenerateSecretName(manifest.Name),
+			Namespace: instance.GetNamespace(),
+		},
+		StringData: map[string]string{
+			"manifest.yaml": string(tempManifestBytes),
+		},
+	}
+
+	foundSecret := &corev1.Secret{}
+	err = r.client.Get(ctx, types.NamespacedName{Name: tempManifestSecret.GetName(), Namespace: tempManifestSecret.GetNamespace()}, foundSecret)
+	if apierrors.IsNotFound(err) {
+		err = r.client.Create(ctx, tempManifestSecret)
+		if err != nil {
+			return errors.Wrap(err, "could not create temp manifest secret")
+		}
+	} else {
+		foundSecret.Data = map[string][]byte{}
+		foundSecret.StringData = map[string]string{
+			"manifest.yaml": string(tempManifestBytes),
+		}
+		err = r.client.Update(ctx, foundSecret)
+		if err != nil {
+			return errors.Wrap(err, "could not update temp manifest secret")
+		}
+	}
+
+	r.log.Debug("Creating variable interpolation extendedJob")
+	varIntExJob, err := r.newExtendedJobTemplateForVariableInterpolation(ctx, foundSecret, variables, jobLabels, secretLabels, instance.GetNamespace())
+	if err != nil && apierrors.IsNotFound(err) {
```

Comment:
Why do you check for `IsNotFound`?
Even if the secrets have not yet been created, afaik you can still create the job. It will keep trying to run until all secrets are ready.

Reply:
This snippet is to upsert temp manifest secret. So it will check `IsNotFound ` and then create or update secret.

Reply:
No, afaict `newExtendedJobTemplateForVariableInterpolation` returns errors if variable secrets don't exist.
See here: https://github.com/cloudfoundry-incubator/cf-operator/pull/147/files/7a9e90535b997b218f8e7ee1f1152ba3b545595c#diff-3c2f267f825b15e51537a7f733f474deR550

Reply:
Sorry. I just mis-replied above section. Yes. What this section want to do is keep trying to run until all secrets are ready. When one secret didn't be created, this func will return `IsNotFound` error to here: https://github.com/cloudfoundry-incubator/cf-operator/blob/76f21dd108fd7ca80ced0ee2115b75e8d390219c/pkg/kube/controllers/boshdeployment/reconciler.go#L199-L203

And reconciler will retry this action.